### PR TITLE
Provide initial support for Mode 9 PIDs

### PR DESCRIPTION
**Pull request by pmundt**
This provides initial support for Mode (Service) 9 PIDs, to the extent that I was able to test them. With this in place, I am able to successfully read the VIN and calibration IDs across a number of OBD-II devices - which should allow issue #42 to be closed out.

### CODE REVIEW

**Comment by alistair23** [comment]
Cool! Thanks for the PR!

I had a quick look and it looks good. Do you mind sharing what you have tested this on? Also can you include a update to the documentation to describe how to use mode 9 and get the VIN?

File: obd/decoders.py [hunk 1]
```diff
@@ -484,3 +489,31 @@ def monitor(messages):
...
+    """ Extract an encoded string from multi-part messages """
+    return functools.partial(decode_encoded_string, length=length)
+
+
+def decode_encoded_string(messages, length):
+    d = messages[0].data[2:]
+
+    # If a bogus string length is returned, discard it
+    if len(d) < length:
+        logger.debug("Invalid string returned. Discarding...")
```
Comment (alistair23) [inline obd/decoders.py]: Is it worth printing the sting here? That way the user can see what was bogus, that should help in debugging

File: obd/obd.py [hunk 2]
```diff
@@ -128,9 +129,11 @@ def __load_commands(self):
             # loop through PIDs bit-array
             for i, bit in enumerate(response.value):
                 if bit:
-
                     mode = get.mode
-                    pid = get.pid + i + 1
+                    if mode == 9:
+                        pid = get.pid + (i - 7) + 1
```
Comment (alistair23) [inline obd/obd.py]: Can you comment why mode 9 is different here?

File: obd/decoders.py [hunk 3]
```diff
@@ -484,3 +489,31 @@ def monitor(messages):
...
+        return None
+
+    # Encoded strings come in bundles of messages with leading null values to
+    # pad out the string to the next full message size. We strip off the
+    # leading null characters here and return the resulting string.
+    while len(d) != length: d.pop(0)
```
Comment (alistair23) [inline obd/decoders.py]: Can't we use lstrip to do this?
https://docs.python.org/3/library/stdtypes.html?highlight=lstrip#str.lstrip

Comment (pmundt) [reply obd/decoders.py]: I tried this with lstrip originally but did not manage to figure out the correct incantation to strip off the null bytes in question - I tried this both with None and with \0, neither of which did the job.

File: obd/obd.py (hunk 2 above)
Comment (pmundt) [reply obd/obd.py]: Unfortunately, no. From the specification itself, the encoding for mode 9 vs mode 1 PIDs should be the same - however, in the cases where I was able to test, there was an off-by-7 on the PID calculation on mode 9 PIDs compared to mode 1. I'm not convinced that this isn't just a bug in how the result is saved off and so the bitmap itself is skewed, but I have not had the opportunity to dig into this more.

File: obd/decoders.py (hunk 1 above)
Comment (pmundt) [reply obd/decoders.py]: We could, but I'm not sure how useful this would be. Multiple PIDs use this decoder with different string formats (e.g. VIN as an ASCII-string, CVN as a hex string, etc.). We could pass in a format string to aid with correctly logging the string output, but this then leads to an uglier and more fragile interface for what is ultimately a corner case.

**Comment by pmundt** [comment]
I have now added the mode 9 PIDs to the command table documentation, with a note about which commands are as of yet still untested.

Regarding my test setup, I have been using a hardware simulator - the [Diamex OBD2 Profi Simulator](https://www.diamex.de/dxshop/Diamex-OBD2-Profi-Simulator-alle-Protokolle), and have tested this with a number of different types of connectivity, including:

- Elm327 BT Dongle
- Elm327 WiFi Dongle
- Direct serial connection to the PI-OBD-HAT

across a fairly wide range of OBD signalling protocols.

In the case of the [PI-OBD-HAT](https://www.diamex.de/dxshop/PI-OBD-HAT-OBD2-Modul-fuer-Raspberry-PI), there are some variations in the command set and the responses that needed to be fixed in order to get python-OBD to work with it, but I will prepare a separate patch series and pull request for this when I have the time.

File: obd/commands.py [hunk 4]
```diff
@@ -8,6 +8,7 @@
 # Copyright 2009 Secons Ltd. (www.obdtester.com)                       #
 # Copyright 2009 Peter J. Creath                                       #
 # Copyright 2016 Brendan Whitfield (brendan-w.com)                     #
+# Copyright 2019 Adaptant Solutions AG                                 #
```
Comment (Ircama) [inline obd/commands.py]: Additional copyright?

File: obd/decoders.py [hunk 5]
```diff
@@ -8,6 +8,7 @@
 # Copyright 2009 Secons Ltd. (www.obdtester.com)                       #
 # Copyright 2009 Peter J. Creath                                       #
 # Copyright 2016 Brendan Whitfield (brendan-w.com)                     #
+# Copyright 2019 Adaptant Solutions AG                                 #
```
Comment (Ircama) [inline obd/decoders.py]: Is additional copyright needed?

File: obd/obd.py [hunk 6]
```diff
@@ -8,6 +8,7 @@
 # Copyright 2009 Secons Ltd. (www.obdtester.com)                       #
 # Copyright 2009 Peter J. Creath                                       #
 # Copyright 2016 Brendan Whitfield (brendan-w.com)                     #
+# Copyright 2019 Adaptant Solutions AG                                 #
```
Comment (Ircama) [inline obd/obd.py]: same (added copyright)

File: obd/decoders.py (hunk 5 above)
Comment (pmundt) [reply obd/decoders.py]: Our contribution, our copyright. What exactly do you perceive to be the issue?

File: obd/decoders.py (hunk 1 above)
Comment (alistair23) [reply obd/decoders.py]: There are lots of different car models, so I think corner cases come up somewhat often. When they do having the string in the log should help debug the issue and will at least improve the quality of the log dumped in a GitHub issue.

File: obd/obd.py (hunk 2 above)
Comment (alistair23) [reply obd/obd.py]: Hmm... That isn't ideal.

That comment should at least be in the code if it doesn't match the spec. If it is a bitmap bug then when that is fixed this can be removed.

File: obd/decoders.py (hunk 3 above)
Comment (alistair23) [reply obd/decoders.py]: Strange, from reading it's documentation this is what it should be used for.

File: obd/decoders.py (hunk 5 above)
Comment (alistair23) [reply obd/decoders.py]: Usually a files copyright is only changed with a major change, like a full re-write. I'm not sure if this is enough of a change to justify adding your copyright

**Comment by alistair23** [comment]
Thanks for the documentation update. It also occurs to me that we should add some unit tests for this so that Travis can ensure it doesn't regress.

File: obd/obd.py (hunk 2 above)
Comment (alistair23) [reply obd/obd.py]: Although thinking about this more we should figure out what is going on here and fix it. I don't really want to add code that expects that the car's responses don't match the spec without good justification. Especially as this looks like it would not work with responses that do match the spec.

It's possible your simulator has a bug in it. I will try to test this series with my set-up and see how it goes.

File: obd/decoders.py (hunk 5 above)
Comment (Ircama) [reply obd/decoders.py]: Contributions are generally made under the same license as the project (unless different agreement). There are [GitHub policies for contributions](https://github.com/github/site-policy/blob/master/Policies/github-terms-of-service.md#6-contributions-under-repository-license).

`git blame` can always be used to view the revision history of lines within a file, and [GitHub *blame*](https://help.github.com/en/articles/tracking-changes-in-a-file) shows the latest line-by-line changes made by respective authors in a file, like here:
https://github.com/brendan-w/python-OBD/blame/8e044f5a797a7ec2ec1341d42250af0bf654ac14/obd/decoders.py

File: docs/Command Tables.md [hunk 7]
```diff
@@ -261,3 +261,25 @@ The return value will be encoded in the same structure as the Mode 03 `GET_DTC`
...
+|PID | Name                         | Description                                        | Response Value        |
+|----|------------------------------|----------------------------------------------------|-----------------------|
+| 00 | PIDS_9A                      | Supported PIDs [01-20]                             | BitArray              |
+| 01 | VIN_MESSAGE_COUNT            | VIN Message Count                                  | Unit.count            |
+| 02 | VIN                          | Get Vehicle Identification Number                  | string                |
+| 03 | CALIBRATION_ID_MESSAGE_COUNT | Calibration ID message count for PID 04            | Unit.count            |
+| 04 | CALIBRATION_ID               | Calibration ID                                     | string                |
+| 05 | CVN_MESSAGE_COUNT            | CVN Message Count for PID 06                       | Unit.count            |
+| 06 | CVN                          | Calibration Verification Numbers                   | hex string            |
+| 07 | PERF_TRACKING_MESSAGE_COUNT  | Performance tracking message count                 | Unit.count            |
```
Comment (alistair23) [inline docs/Command Tables.md]: PERF_TRACKING_MESSAGE_COUNT and below are not included in the source

**Comment by alistair23** [comment]
This is the log that I get:

> [obd.obd] Sending command: b'0902': Get Vehicle Identification Number
> [obd.elm327] write: b'0902\r'
> [obd.elm327] read: b'7E8 10 14 49 02 01 57 50 30 \r7E8 21 5A 5A 5A 39 39 5A 54 \r7E8 22 53 33 39 32 31 32 34 \r\r>'
> [obd.OBDCommand] Message was shorter than expected (20<22). Padded message: bytearray(b'I\x02\x01WP0ZZZ99ZTS392124\x00\x00')
> bytearray(b'0ZZZ99ZTS392124\x00\x00')

With this diff I get some better numbers:

```diff
diff --git a/obd/commands.py b/obd/commands.py
index ff8dd36..50c16bf 100644
--- a/obd/commands.py
+++ b/obd/commands.py
@@ -283,13 +283,13 @@ __mode7__ = [

 __mode9__ = [
     #                      name                             description                            cmd     bytes       decoder       ECU        fast
-    OBDCommand("PIDS_9A"                    , "Supported PIDs [01-20]"                            , b"0900",  7, pid,                ECU.ALL,     True),
+    OBDCommand("PIDS_9A"                    , "Supported PIDs [01-20]"                            , b"0900",  6, pid,                ECU.ALL,     True),
     OBDCommand("VIN_MESSAGE_COUNT"          , "VIN Message Count"                                 , b"0901",  3, count,              ECU.ENGINE,  True),
-    OBDCommand("VIN"                        , "Get Vehicle Identification Number"                 , b"0902", 22, encoded_string(17), ECU.ENGINE,  True),
+    OBDCommand("VIN"                        , "Get Vehicle Identification Number"                 , b"0902", 20, encoded_string(17), ECU.ENGINE,  True),
     OBDCommand("CALIBRATION_ID_MESSAGE_COUNT","Calibration ID message count for PID 04"           , b"0903",  3, count,              ECU.ALL,     True),
     OBDCommand("CALIBRATION_ID"             , "Calibration ID"                                    , b"0904", 18, encoded_string(16), ECU.ALL,     True),
     OBDCommand("CVN_MESSAGE_COUNT"          , "CVN Message Count for PID 06"                      , b"0905",  3, count,              ECU.ALL,     True),
-    OBDCommand("CVN"                        , "Calibration Verification Numbers"                  , b"0906", 10, cvn,                ECU.ALL,     True),
+    OBDCommand("CVN"                        , "Calibration Verification Numbers"                  , b"0906", 18, cvn,                ECU.ALL,     True),

 #
 # NOTE: The following are untested
diff --git a/obd/decoders.py b/obd/decoders.py
index 05062ea..9c02b0e 100644
--- a/obd/decoders.py
+++ b/obd/decoders.py
@@ -504,11 +504,6 @@ def decode_encoded_string(messages, length):
         logger.debug("Invalid string returned. Discarding...")
         return None

-    # Encoded strings come in bundles of messages with leading null values to
-    # pad out the string to the next full message size. We strip off the
-    # leading null characters here and return the resulting string.
-    while len(d) != length: d.pop(0)
-
     return d
```

> [obd.obd] Sending command: b'0902': Get Vehicle Identification Number
> [obd.elm327] write: b'0902\r'
> [obd.elm327] read: b'7E8 10 14 49 02 01 57 50 30 \r7E8 21 5A 5A 5A 39 39 5A 54 \r7E8 22 53 33 39 32 31 32 34 \r\r>'
> bytearray(b'\x01WP0ZZZ99ZTS392124')

This is a great start, but I think it's going to need some more testing to match up. The best bet is to probably write some unit tests based on the spec and then we can use them as a starting point. Then we just need to test on more cars/simulators.

**Comment by Ircama** [comment]
@alistair23, @pmundt Could you provide full traces? I can help emulating the workflow.

**Comment by alistair23** [comment]
Yep, attached [vin_test.log](https://github.com/brendan-w/python-OBD/files/3337353/vin_test.log)

File: obd/commands.py [hunk 8]
```diff
@@ -279,6 +280,27 @@
     OBDCommand("GET_CURRENT_DTC", "Get DTCs from the current/last driving cycle", b"07", 0, dtc, ECU.ALL, False),
 ]
 
+
+__mode9__ = [
+    #                      name                             description                            cmd     bytes       decoder       ECU        fast
+    OBDCommand("PIDS_9A"                    , "Supported PIDs [01-20]"                            , b"0900",  7, pid,                ECU.ALL,     True),
+    OBDCommand("VIN_MESSAGE_COUNT"          , "VIN Message Count"                                 , b"0901",  3, count,              ECU.ENGINE,  True),
+    OBDCommand("VIN"                        , "Get Vehicle Identification Number"                 , b"0902", 22, encoded_string(17), ECU.ENGINE,  True),
```
Comment (Ircama) [inline obd/commands.py]: Is "`Get `" needed?

Comment (BedoM) [reply obd/commands.py]: How can the subject progress?

Can we help?

**Comment by alistair23** [comment]
@BedoM All the comments in the PR need to be addressed, including setting up pytests. If you are up for you you can take this PR, fix the issues and add tests then it should be ok to merge. Feel free to create a new PR with the work, feel free to ask for any help you need, probably the best place is in the issue you created.

**Comment by BedoM** [comment]
thank you for your answers, but, i think not have the skills for this.

I see if i have the skills..

**Comment by alistair23** [comment]
I just created a PR based on this work that adds support for reading the VIN.

It address all of the original comments. It would be great if anyone interested can test it in their setup to make sure it works.

https://github.com/brendan-w/python-OBD/pull/171

**Comment by alistair23** [comment]
Closing this PR as it outdated.
